File: lib/domain/cultivation_plan/gateways/plan_allocation_adjust_gateway.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Gateways
      class PlanAllocationAdjustGateway
        # @raise [Domain::CultivationPlan::Errors::AdjustExecutionError] when agrr adjust execution fails
        def adjust(current_allocation:, moves:, fields:, crops:, weather_data:, planning_start:, planning_end:, interaction_rules: nil, objective: "maximize_profit", max_time: nil, enable_parallel: false)
          raise NotImplementedError, "#{self.class}##{__method__}"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/gateways/plan_allocation_adjust_gateway.rs
//! Ruby: `Domain::CultivationPlan::Gateways::PlanAllocationAdjustGateway`

use crate::cultivation_plan::errors::AdjustExecutionError;
use serde_json::Value;
use time::Date;

pub trait PlanAllocationAdjustGateway: Send + Sync {
    #[allow(clippy::too_many_arguments)]
    fn adjust(
        &self,
        current_allocation: &Value,
        moves: &[Value],
        fields: &[Value],
        crops: &[Value],
        weather_data: &Value,
        planning_start: Date,
        planning_end: Date,
        interaction_rules: Option<&Value>,
        objective: &str,
        max_time: Option<i64>,
        enable_parallel: bool,
    ) -> Result<Value, AdjustExecutionError>;
}
